src/linear-types/Control/Distributed/Spark/Safe/Closure.hs

```diff
@@ -0,0 +1,303 @@
+  S.deleteLocalRef objptr
+  S.deleteLocalRef args'
+  upcast <$> refl (f a')
+    -- fmap upcast . refl =<< return . f =<< reif . unsafeCast =<< S.getObjectArrayElement args 0
+  where
+    reif = reify :: J (Interp a) %1 -> LIO.IO (J (Interp a), Ur a)
+    refl = reflect :: b -> LIO.IO (J (Interp b))
+
+tripleDict :: Dict c1 -> Dict c2 -> Dict c3 -> Dict (c1, c2, c3)
+tripleDict Dict Dict Dict = Dict
+
+closFun2
+  :: forall a b c.
+    Dict (Reify a, Reify b, Reflect c)
+     -> (a -> b -> c)
+     -> JObjectArray
+  %1 -> LIO.IO JObject
+closFun2 Dict f args = Control.Functor.Linear.do
+    (args', a) <- second unsafeCast <$> S.getObjectArrayElement args 0
+    (args'', b) <- second unsafeCast <$> S.getObjectArrayElement args' 1
+    (aptr, Ur a') <- reifA a
+    (bptr, Ur b') <- reifB b
+    S.deleteLocalRef aptr
+    S.deleteLocalRef bptr
+    S.deleteLocalRef args''
+    upcast <$> reflC (f a' b')
+  where
+    reifA = reify :: J (Interp a) %1 -> LIO.IO (J (Interp a), Ur a)
+    reifB = reify :: J (Interp b) %1 -> LIO.IO (J (Interp b), Ur b)
+    reflC = reflect :: c -> LIO.IO (J (Interp c))
+    second :: (n %1 -> o) -> (m, n) %1 -> (m, o)
+    second g (a, b) = (a, g b) 
+
+clos2bs :: Typeable a => Closure a %1-> ByteString
+clos2bs = Unsafe.toLinear $ LBS.toStrict Prelude.. encode
+
+bs2clos :: Typeable a => ByteString %1-> Closure a
+bs2clos = Unsafe.toLinear $ decode Prelude.. LBS.fromStrict
+
+-- TODO: These two instances should probably go in `jvm` instead of here
+instance (Static (JUnsafe.Reify a), Reify a) => Static (Reify a) where
+  closureDict = closureDict
+
+instance (Static (JUnsafe.Reflect a), Reflect a) => Static (Reflect a) where
+  closureDict = closureDict
+
+-- | Like 'Interp', but parameterized by the target arity of the 'Fun' instance.
+type family InterpWithArity (n :: Nat) (a :: Type) :: JType
+
+-- | A @ReifyFun n a@ constraint states that @a@ is a function type of arity
+-- @n@. That is, a value of this function type can be made from a @JFun{n}@.
+class ReifyFun n a where
+  -- | Like 'reify', but specialized to reifying functions at a given arity.
+  reifyFun :: MonadIO m => Sing n -> J (InterpWithArity n a) %1 -> m (J (InterpWithArity n a), Ur a)
+
+-- TODO define instances for 'ReifyFun'.
+
+-- | A @ReflectFun n a@ constraint states that @a@ is a function type of arity
+-- @n@. That is, a @JFun{n}@ can be made from any value of this function type.
+class ReflectFun n a where
+  -- | Like 'reflect', but specialized to reflecting functions at a given arity.
+  reflectFun :: MonadIO m => Sing n -> Closure a -> m (J (InterpWithArity n a))
+
+-- TODO No Static (Fun (a -> b)) instances yet.
+
+type instance InterpWithArity 1 (a -> b) = JFun1 (Interp a) (Interp b)
+type JFun1 a b = 'Iface "org.apache.spark.api.java.function.Function" <> [a, b]
+
+instance ( Static (Reify a)
+         , Static (Reflect b)
+         , Typeable a
+         , Typeable b
+         ) =>
+         ReflectFun 1 (a -> b) where
```

Comment:
Strictly speaking `a` here can be instantiated to `J ty`, while `a -> b` is not a linear function. However, `a -> b` is pure, and we can do nothing with a java reference in a pure context, not even deleting it.

Perhaps we don't need to worry about it here, but then `MapPartitionsFunction` below does allow for side effects and probably should be linear.

Reply:
This could be regarded also as a flaw in inline-java's `Reify` class.
```
reify :: MonadIO m => J (Interp a) %1-> m (J (Interp a), Ur a)
```
We are allowing non-linear references to be returned.

In that light, I don't think we need to do anything to address the problem in sparkle.